### PR TITLE
[Sensor] Initial implementation of the Generic Sensor API

### PR DESCRIPTION
**Pull request by jimmy-huang**
Initial implementation for the W3C Generic Sensor API and add
support for the Accelerometer and Gyroscope sensors using
the BMI160 inertia sensor built-in on the Arduino 101,
the current implementation is polling from the ARC and channeling
the data back to X86.

Also adding two sample JS code testing both APIs.

Signed-off-by: Jimmy Huang <jimmy.huang@intel.com>

### CODE REVIEW

File: arc/src/main.c [hunk 1]
```diff
@@ -322,6 +324,7 @@ static void handle_glcd(struct zjs_ipm_message* msg)
             glcd = device_get_binding(GROVE_LCD_NAME);
 
             if (!glcd) {
+                PRINT("failed to initialize Grove LCD\n");
```
Comment (kenchris) [inline arc/src/main.c]: Seems unrelated

File: src/zjs_sensor.c [hunk 2]
```diff
@@ -0,0 +1,512 @@
...
+        if (jerry_value_has_error_flag(rval)) {
+            PRINT("zjs_sensor_set_state: error calling onstatechange\n");
+        }
+        jerry_release_value(rval);
+    }
+    jerry_release_value(func);
+
+    if (old_state == SENSOR_STATE_ACTIVATING &&
+        state == SENSOR_STATE_ACTIVATED) {
+        func = zjs_get_property(obj, "onactivate");
```
Comment (kenchris) [inline src/zjs_sensor.c]: Great that you are actively following the spec, this was just added two days ago :-)

https://github.com/w3c/sensors/commit/a85830bef0566b21176ac5af1ba0b1d8241b06ca

File: src/zjs_sensor.c [hunk 3]
```diff
@@ -0,0 +1,512 @@
...
+#include "zjs_sensor.h"
+#include "zjs_callbacks.h"
+#include "zjs_ipm.h"
+#include "zjs_util.h"
+
+#define ZJS_SENSOR_TIMEOUT_TICKS                      500
+
+static struct nano_sem sensor_sem;
+
+enum sensor_state {
```
Comment (kenchris) [inline src/zjs_sensor.c]: I believe these might become internals: Check

https://github.com/w3c/sensors/issues/104#issuecomment-248316955

(that part is not fixed in spec yet, ie the checkbox isnt checked yet)

Comment (kenchris) [inline src/zjs_sensor.c]: We discussed these changes back in the W3C TPAC meeting in Lisbon and they are slowly making it into the spec

File: src/zjs_sensor.c [hunk 4]
```diff
@@ -0,0 +1,512 @@
...
+
+        double option_freq;
+        if (zjs_obj_get_double(options, "frequency", &option_freq)) {
+            // TODO: figure out a list of frequencies we can support
+            // and have Zephyr trigger value changes instead of
+            // polling on the ARC.
+            // For now, frequency is ignored,  we just report new event
+            // as soon as we detect a value change.
+            if (option_freq <= 0) {
+                PRINT("zjs_sensor_create: invalid frequency, default to 50hz\n");
```
Comment (kenchris) [inline src/zjs_sensor.c]: defaulting

File: src/zjs_sensor.c [hunk 5]
```diff
@@ -0,0 +1,512 @@
...
+            }
+        } else {
+            PRINT("zjs_sensor_create: frequency not found, default to 50hz\n");
+        }
+
+        if (channel == SENSOR_CHAN_ACCEL_ANY) {
+            bool option_gravity;
+            if (zjs_obj_get_boolean(options, "includeGravity", &option_gravity) &&
+                option_gravity) {
+                // TODO: find out if BMI160 can be configured to include gravity
```
Comment (kenchris) [inline src/zjs_sensor.c]: I guess the math to add it manually wouldn't be too hard

File: src/zjs_ipm.h [hunk 6]
```diff
@@ -95,6 +124,15 @@ typedef struct zjs_ipm_message {
             uint8_t color_b;
             void *buffer;
         } glcd;
+#endif
+
+#ifdef BUILD_MODULE_SENSOR
+        struct sensor_data {
```
Comment (kenchris) [inline src/zjs_ipm.h]: For external sensors (ie DHT22 etc) we would probably need something like "controller" here so that I can do

new Hygrometer( { controller: "dht22" } ) which would hook up to the right Zephyr driver

File: scripts/analyze.sh [hunk 7]
```diff
@@ -104,6 +104,11 @@ if [ $? -eq 0 ] && [[ $MODULE != *"BUILD_MODULE_BUFFER"* ]]; then
     >&2 echo Using module: Buffer
     MODULES+=" -DBUILD_MODULE_BUFFER"
 fi
+sensor=$(grep -E AccelerometerSensor\|Gyroscope  $SCRIPT)
```
Comment (kenchris) [inline scripts/analyze.sh]: I thought the Sensor part was being removed... I will ask Riju to look at this PR

File: samples/Accelerometer.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
+// Copyright (c) 2016, Intel Corporation.
+
+// Test code to use the AccelerometerSensor (subclass of Generic Sensor) API
+// to communicate with the BMI160 inertia sensor on the Arduino 101
+// and obtaining information about acceleration applied to the X, Y and Z axis
+console.log("Acclerometer test...");
```
Comment (kenchris) [inline samples/Accelerometer.js]: spelling error

File: samples/Gyroscope.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
+// Copyright (c) 2016, Intel Corporation.
+
+// Test code to use the Gyroscope (subclass of Generic Sensor) API
+// to communicate with the BMI160 inertia sensor on the Arduino 101
+// and monitor the rate of rotation around the the X, Y and Z axis
+print("Gyroscope test...");
```
Comment (kenchris) [inline samples/Gyroscope.js]: Riju, do you have tests from Blink that we can reuse?

File: src/zjs_sensor.c [hunk 10]
```diff
@@ -0,0 +1,512 @@
...
+    jerry_value_t x_val = jerry_create_number(x);
+    jerry_value_t y_val = jerry_create_number(y);
+    jerry_value_t z_val = jerry_create_number(z);
+    jerry_value_t reading_obj = jerry_create_object();
+    if (channel == SENSOR_CHAN_ACCEL_ANY) {
+        zjs_set_property(reading_obj, "accelerationX", x_val);
+        zjs_set_property(reading_obj, "accelerationY", y_val);
+        zjs_set_property(reading_obj, "accelerationZ", z_val);
+    } else if (channel == SENSOR_CHAN_GYRO_ANY) {
+        zjs_set_property(reading_obj, "rotationRateX", x_val);
```
Comment (riju) [inline src/zjs_sensor.c]: same here

File: samples/Accelerometer.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+
+var sensor = new AccelerometerSensor({
+    includeGravity: false, // true is not supported, will throw error
+    frequency: 60
+});
+
+sensor.onchange = function(event) {
+    console.log("acceleration (m/s^2): " +
+          " x=" + event.reading.accelerationX +
```
Comment (riju) [inline samples/Accelerometer.js]: accelerationX -> x , changed recently
https://w3c.github.io/accelerometer/#accelerometerreading

File: samples/Gyroscope.js [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+          " y=" + event.reading.rotationRateY +
+          " z=" + event.reading.rotationRateZ);
+};
+
+sensor.onstatechange = function(event) {
+    console.log("state: " + event);
+};
+
+sensor.onerrorchange = function(error) {
+    console.log("error: " + error);
```
Comment (riju) [inline samples/Gyroscope.js]: nit : sensor.onerror = event => console.log(event.error.name, event.error.message);

File: scripts/analyze.sh (hunk 7 above)
Comment (riju) [inline scripts/analyze.sh]: Yes, we have remoed the "sensor" part. https://w3c.github.io/accelerometer/#accelerometer

File: samples/Gyroscope.js (hunk 9 above)
Comment (riju) [inline samples/Gyroscope.js]: We have the AmbientLightSensor tests here -> https://cs.chromium.org/chromium/src/third_party/WebKit/LayoutTests/sensor/ambient-light-sensor.html?q=ambient&sq=package:chromium&l=1

Gyro and Accelerometer test cases will be upstreamed along with their implementation soon.

File: samples/Accelerometer.js [hunk 13]
```diff
@@ -0,0 +1,29 @@
+// Copyright (c) 2016, Intel Corporation.
+
+// Test code to use the AccelerometerSensor (subclass of Generic Sensor) API
+// to communicate with the BMI160 inertia sensor on the Arduino 101
+// and obtaining information about acceleration applied to the X, Y and Z axis
+console.log("Acclerometer test...");
+
+
+var sensor = new AccelerometerSensor({
```
Comment (riju) [inline samples/Accelerometer.js]: AccelerometerSensor -> Accelerometer
Just a heads up, very soon, we will split the Accelerometer interface to 2 interfaces, one for Accelerometer and another for LinearAccelerometer. In that case includeGravity will not be needed. You may want to change this when it is reflected in the spec.

File: samples/Accelerometer.js [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+          " x=" + event.reading.accelerationX +
+          " y=" + event.reading.accelerationY +
+          " z=" + event.reading.accelerationZ);
+};
+
+sensor.onstatechange = function(event) {
+    console.log("state: " + event);
+};
+
+sensor.onerrorchange = function(error) {
```
Comment (riju) [inline samples/Accelerometer.js]: sensor.onerror = event => console.log(event.error.name, event.error.message);

File: src/zjs_ipm.h [hunk 15]
```diff
@@ -95,6 +124,15 @@ typedef struct zjs_ipm_message {
...
+            enum sensor_channel channel;
+            uint32_t frequency;
+            double value[3];
```
Comment (riju) [inline src/zjs_ipm.h]: Would you prefer it to split this struct into 2 ? Like we have for SensorReading (frequency&timestamp) as base struct, and say concreteSensor's reading (which contains the actual value) like https://w3c.github.io/gyroscope/#gyroscopereading. 
Also no timestamp ? like performance.now() equivalent .

When you implement the next sensor, say AmbientLightSensor , there is only 1 double for "value".
If you implement Uncalibrated Magnetometer or Uncalibrated Gyro, then double value[6].

File: src/zjs_sensor.c [hunk 16]
```diff
@@ -0,0 +1,512 @@
...
+    SENSOR_STATE_ACTIVATING,
+    SENSOR_STATE_ACTIVATED,
+    SENSOR_STATE_ERRORED,
+};
+
+typedef struct sensor_handle {
+    int32_t id;
+    enum sensor_channel channel;
+    enum sensor_state state;
+    double sensor_value[3];
```
Comment (riju) [inline src/zjs_sensor.c]: Fine for now, but not generic enough. See https://github.com/01org/zephyr.js/pull/353/commits/294d36b85640c752a6840df72dbe3073278b9aa7#r86344326

File: arc/src/main.c (hunk 1 above)
Comment (jimmy-huang) [inline arc/src/main.c]: Yes, I forgot to add it to the glcd when I am implementing sensor, so I am thinking to include it as part of the patch, but I'll remove it and throw in another patch.

File: scripts/analyze.sh (hunk 7 above)
Comment (jimmy-huang) [inline scripts/analyze.sh]: Ok, I'll update accordingly, I was wondering that too since it's odd that the one has, but one does not.

File: samples/Gyroscope.js [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+// to communicate with the BMI160 inertia sensor on the Arduino 101
+// and monitor the rate of rotation around the the X, Y and Z axis
+print("Gyroscope test...");
+
+
+var sensor = new Gyroscope();
+
+sensor.onchange = function(event) {
+    print("rotation (rad/s): " +
+          " x=" + event.reading.rotationRateX +
```
Comment (jimmy-huang) [inline samples/Gyroscope.js]: I'll update accordingly.

File: samples/Gyroscope.js (hunk 12 above)
Comment (jimmy-huang) [inline samples/Gyroscope.js]: This is a limitation of the jerryscript engine, it doesn't support this syntax, it's ES6 right?  It only supports ES5 for now.

File: samples/Accelerometer.js (hunk 13 above)
Comment (jimmy-huang) [inline samples/Accelerometer.js]: Ok, I'll keep that in mind, thanks for the heads up.

File: samples/Accelerometer.js (hunk 14 above)
Comment (jimmy-huang) [inline samples/Accelerometer.js]: See my comment above.

File: src/zjs_ipm.h [hunk 18]
```diff
@@ -59,7 +77,16 @@ enum {
...
 #define TYPE_GLCD_GET_INPUT_STATE                          0x002B
+#endif
 
+// SENSOR
+#ifdef BUILD_MODULE_SENSOR
+#define TYPE_SENSOR_INIT                                   0x0030
+#define TYPE_SENSOR_START                                  0x0031
+#define TYPE_SENSOR_STOP                                   0x0032
+#define TYPE_SENSOR_EVENT_STATE_CHANGE                     0x0033
+#define TYPE_SENSOR_EVENT_VALUE_CHANGE                     0x0034
```
Comment (jimmy-huang) [inline src/zjs_ipm.h]: that's fine with me

File: src/zjs_ipm.h (hunk 15 above)
Comment (jimmy-huang) [inline src/zjs_ipm.h]: ok yeah, this is the internal implementation, and only specific to Arduino 101 where we have to channel the data, so we can fix it up however we want, but that's a good idea.

File: src/zjs_sensor.c (hunk 5 above)
Comment (jimmy-huang) [inline src/zjs_sensor.c]: that's one possible way to do it

File: samples/Gyroscope.js (hunk 12 above)
Comment (riju) [inline samples/Gyroscope.js]: good to know that. iirc, its better to use event, instead of error object. not sure, right now.

```
sensor.onerror = function(event) {
   console.log(event.error.name, event.error.message);
```

File: src/zjs_ipm.h (hunk 15 above)
Comment (jimmy-huang) [inline src/zjs_ipm.h]: @riju I updated patch to use a union to hold the reading values depending on the sensor type

File: arc/src/main.c [hunk 19]
```diff
@@ -388,6 +388,250 @@ static void handle_glcd(struct zjs_ipm_message* msg)
     ipm_send_reply(msg);
 }
 
+#ifndef DEBUG_BUILD
```
Comment (grgustaf) [inline arc/src/main.c]: I think you meant #ifdef? This is used only in DEBUG_BUILD below AFAICT. Bad Jimmy, didn't build with DEBUG! :)

File: arc/src/main.c [hunk 20]
```diff
@@ -388,6 +388,250 @@ static void handle_glcd(struct zjs_ipm_message* msg)
...
+    msg.type = TYPE_SENSOR_EVENT_READING_CHANGE;
+    msg.flags = 0;
+    msg.user_data = NULL;
+    msg.error_code = ERROR_IPM_NONE;
+    msg.data.sensor.channel = channel;
+    memcpy(&msg.data.sensor.reading, &reading, sizeof(union sensor_reading));
+    zjs_ipm_send(MSG_ID_SENSOR, &msg);
+}
+
+#define ABS(x) ((x) > 0) ? (x) : -(x)
```
Comment (grgustaf) [inline arc/src/main.c]: &gt;= perhaps?

File: arc/src/main.c [hunk 21]
```diff
@@ -388,6 +388,250 @@ static void handle_glcd(struct zjs_ipm_message* msg)
...
+            val2 = val->val2;
+        }
+
+        if (val1 > 0 || (val1 == 0 && val2 > 0)) {
+            result = val1 + (double)val2 * 0.000001;
+        } else if (val1 == 0 && val2 < 0) {
+            result = (double)val2 * (-0.000001);
+        } else {
+            result = val1 + (double)val2 * (-0.000001);
+        }
```
Comment (grgustaf) [inline arc/src/main.c]: What happens in this block if val1 > 0 and val2 < 0, doesn't it do the wrong thing?

It kind of seems like this logic could be simplified to be more obviously correct. I'm still not sure what it's doing but will wait for your response to this before digging deeper. :)

File: arc/src/main.c [hunk 22]
```diff
@@ -388,6 +388,250 @@ static void handle_glcd(struct zjs_ipm_message* msg)
...
+        PRINT("Cannot read accelerometer channels.\n");
+        return;
+    }
+
+    dval[0] = convert_sensor_value(&val[0]);
+    dval[1] = convert_sensor_value(&val[1]);
+    dval[2] = convert_sensor_value(&val[2]);
+
+    if (dval[0] == 0 && dval[1] == 0 && dval[2] == 0) {
+        // FIXME: BUG? why sometimes it reports 0, 0, 0 on all axis
```
Comment (grgustaf) [inline arc/src/main.c]: axis -> axes.  Did you report that to Zephyr folks? Maybe open a JIRA?

File: arc/src/main.c [hunk 23]
```diff
@@ -388,6 +388,250 @@ static void handle_glcd(struct zjs_ipm_message* msg)
...
+        return;
+    }
+
+    // set slope threshold to 0.1G (0.1 * 9.80665 = 4.903325 m/s^2)
+    double threshold = 0.980665;
```
Comment (grgustaf) [inline arc/src/main.c]: Now that's some accurate gravity, right there. Is that calibrated to the second floor of JF1?

File: arc/src/main.c [hunk 24]
```diff
@@ -388,6 +388,250 @@ static void handle_glcd(struct zjs_ipm_message* msg)
...
+        return;
+    }
+
+    dval[0] = convert_sensor_value(&val[0]);
+    dval[1] = convert_sensor_value(&val[1]);
+    dval[2] = convert_sensor_value(&val[2]);
+
+    if (ABS(dval[0] - gyro_last_value[0]) > 0 ||
+        ABS(dval[1] - gyro_last_value[1]) > 0 ||
+        ABS(dval[2] - gyro_last_value[2]) > 0) {
```
Comment (grgustaf) [inline arc/src/main.c]: If your threshold for difference is 0, then isn't this the same as dval[0] != gyro_last_value[0]?

File: src/zjs_ipm.h [hunk 25]
```diff
@@ -4,15 +4,27 @@
...
+#endif
+#ifdef BUILD_MODULE_I2C
 #define MSG_ID_I2C                                         0x02
+#endif
+#ifdef BUILD_MODULE_GROVE_LCD
 #define MSG_ID_GLCD                                        0x03
+#endif
+#ifdef BUILD_MODULE_SENSOR
+#define MSG_ID_SENSOR                                      0x04
+#endif
```
Comment (grgustaf) [inline src/zjs_ipm.h]: I think you can go ahead and define all of these all the time, without the ifdefs. And actually I think elsewhere you use them without being in an ifdef so that would break the build anyway. Maybe review your usage of MSG_ID_* everywhere to make sure it makes sense.

File: src/zjs_ipm.h [hunk 26]
```diff
@@ -26,27 +38,33 @@ enum {
...
 // AIO
+#ifdef BUILD_MODULE_AIO
 #define TYPE_AIO_OPEN                                      0x0000
 #define TYPE_AIO_PIN_READ                                  0x0001
 #define TYPE_AIO_PIN_ABORT                                 0x0002
 #define TYPE_AIO_PIN_CLOSE                                 0x0003
 #define TYPE_AIO_PIN_SUBSCRIBE                             0x0004
 #define TYPE_AIO_PIN_UNSUBSCRIBE                           0x0005
 #define TYPE_AIO_PIN_EVENT_VALUE_CHANGE                    0x0006
+#endif
```
Comment (grgustaf) [inline src/zjs_ipm.h]: Defining macros like this that are never used (if the corresponding module isn't built) doesn't really hurt anything, so I doubt we need these #ifdefs?

File: src/zjs_ipm.h [hunk 27]
```diff
@@ -69,13 +96,14 @@ typedef struct zjs_ipm_message {
     uint32_t error_code;
 
     union {
-        // AIO
+#ifdef BUILD_MODULE_AIO
```
Comment (grgustaf) [inline src/zjs_ipm.h]: Even defining structs that don't get used shouldn't really hurt anything. A few microseconds faster built time but more clutter in the code.

File: src/zjs_sensor.c [hunk 28]
```diff
@@ -0,0 +1,521 @@
...
+
+#define ZJS_SENSOR_TIMEOUT_TICKS                      500
+
+static struct nano_sem sensor_sem;
+
+enum sensor_state {
+    SENSOR_STATE_IDLE,
+    SENSOR_STATE_ACTIVATING,
+    SENSOR_STATE_ACTIVATED,
+    SENSOR_STATE_ERRORED,
```
Comment (grgustaf) [inline src/zjs_sensor.c]: Is this non-word Errored something you're inheriting from elsewhere? :)

**Comment by grgustaf** [comment]
Trying out Accelerometer.js, when my board is sitting flat it reports something like:
acceleration (m/s^2):  x=-0.92958 y=0.25773799999999997 z=11.353628

We would expect x=0, y=0, z=9.81; so y is understandable but x and z seem surprisingly off I think?

**Comment by grgustaf** [comment]
It would be very nice if the output of the sample could be printed with a fixed width, so the y and z values line up as they're spewing by, then you could watch them go up and down without having to pause the stream.

File: src/zjs_sensor.c [hunk 29]
```diff
@@ -0,0 +1,521 @@
...
+        zjs_sensor_update_reading(handle->sensor_obj,
+                                  handle->channel,
+                                  handle->reading);
+    }
+}
+
+static void zjs_sensor_signal_callbacks(sensor_handle_t *handle,
+                                        union sensor_reading reading)
+{
+    if (!handle)
```
Comment (grgustaf) [inline src/zjs_sensor.c]: This isn't needed because your loop will be skipped if it's NULL.

**Comment by grgustaf** [comment]
I can confirm the calibration made the flat on desk case more accurate. If I tip it up on short end or long end it shows x=-8.3 though or y=-8.3, where i'd expect -9.81. But flipped the opposite way where gravity is positive they are around 9.8. Not sure if it can get better, especially without a dynamic user involved calibration?

File: arc/src/main.c (hunk 19 above)
Comment (jimmy-huang) [inline arc/src/main.c]: yeah, lol

File: arc/src/main.c (hunk 20 above)
Comment (jimmy-huang) [inline arc/src/main.c]: ok, i'll change it

File: arc/src/main.c (hunk 21 above)
Comment (jimmy-huang) [inline arc/src/main.c]: if val1 > 0, then it falls into the first case, and if val1 > 0, then val2 can't be < 0 as the sign bit is in val1.  But anyway, this is taken from the the zephyr sample how they checked.

File: arc/src/main.c (hunk 22 above)
Comment (jimmy-huang) [inline arc/src/main.c]: I think this will be fixed with the trigger, that's why I wasn't looking into this.

File: arc/src/main.c (hunk 24 above)
Comment (jimmy-huang) [inline arc/src/main.c]: yeah, i was going to add a threshold here intead of 0, but I don't really know what value to put in for now, so i left it as 0.  I'll update it.

File: src/zjs_ipm.h (hunk 25 above)
Comment (jimmy-huang) [inline src/zjs_ipm.h]: ok, i'll revert them.

**Comment by grgustaf** [comment]
Jimmy and I found that the last two commits segfault if you try to run AIO.js. If you run the BTGrove demo, it does the same once you start to connect.

File: samples/Accelerometer.js [hunk 30]
```diff
@@ -0,0 +1,30 @@
...
+// and obtaining information about acceleration applied to the X, Y and Z axis
+console.log("Accelerometer test...");
+
+
+var sensor = new Accelerometer({
+    includeGravity: false, // true is not supported, will throw error
+    frequency: 50
+});
+
+sensor.onchange = function(event) {
```
Comment (riju) [inline samples/Accelerometer.js]: @jimmy-huang : Some changes were made for the [Sensor Reading](https://w3c.github.io/sensors/#sensor-reading)
Now
```js
accleromterSensor.onchange = function () {
console.log("acceleration (m/s^2): " +
        " x=" + accleromterSensor.reading.x +
        " y=" + accleromterSensor.reading.y +
        " z=" + accleromterSensor.reading.z);
}
```
just a heads up, you can make this changes in subsequent PR also, upto you.

File: samples/Accelerometer.js [hunk 31]
```diff
@@ -0,0 +1,30 @@
...
+    console.log("acceleration (m/s^2): " +
+                " x=" + event.reading.x +
+                " y=" + event.reading.y +
+                " z=" + event.reading.z);
+};
+
+sensor.onstatechange = function(event) {
```
Comment (riju) [inline samples/Accelerometer.js]: onstatechange is also removed

File: samples/Gyroscope.js [hunk 32]
```diff
@@ -0,0 +1,27 @@
...
+// to communicate with the BMI160 inertia sensor on the Arduino 101
+// and monitor the rate of rotation around the the X, Y and Z axis
+console.log("Gyroscope test...");
+
+
+var sensor = new Gyroscope();
+
+sensor.onchange = function(event) {
+    console.log("rotation (rad/s): " +
+                " x=" + event.reading.x +
```
Comment (riju) [inline samples/Gyroscope.js]: Please see the comment above for accelerometer

**Comment by jimmy-huang** [comment]
this needs to be rebased once #415 is merged

**Comment by jimmy-huang** [comment]
@grgustaf, updated pr and rebased, please review
